### PR TITLE
 gcp: check quotas for GCP before creating cluster 

### PR DESCRIPTION
**Pull request by abhinavdahiya**
### loading the limits for quotas
[how to manage quotas](https://cloud.google.com/service-usage/docs/manage-quota) provides information how how to list the limits for various quotas.

Here we use the API documented in [reference](https://cloud.google.com/service-usage/docs/reference/rest/v1beta1/services.consumerQuotaMetrics/list)

For each [CosumerQuotaMetric](https://cloud.google.com/service-usage/docs/reference/rest/v1beta1/services.consumerQuotaMetrics#ConsumerQuotaMetric) we have a list of [QuotaBuckets](https://cloud.google.com/service-usage/docs/reference/rest/v1beta1/services.consumerQuotaMetrics.limits#ConsumerQuotaLimit.QuotaBucket) in order of increasing constraint based on dimension...
no dimension => `global`
dimension keyed on `region` => corresponding region.

We store a limits per Quota bucket so certain limits are global while some are both global and regional.

### loading the usages for quotas

We use the API document in [reference](https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.timeSeries/list) to list timseries of type `serviceruntime.googleapis.com/quota/allocation/usage` and resource type `consumer_quota` for a specific project in time internal [now-1hr, now-5min] .

We use the first Point for each metric as the usage as the values sorted in latest to oldest, and defaulting to zero usage if not points exist.

### combining quota limits with usages

The limits are constrained on regions, while the usages are defined on Zones, so the merge functions aggregates the usages across zones to the common region. Otherwise we match the (service, name, location) across the limits and usages to create a concise information of various quotas.

### Defining resource constraints for create cluster

The commit https://github.com/openshift/installer/pull/3819/commits/fa77d0e21fe6a36aa59b4497edabe9650c39b368 should be easy to follow to figure out how we generate a list of resource requirement based on the install-config, list of controlPlane machine objects and a list of compute MachineSet object.

### CODE REVIEW

**Comment by patrickdillon** [comment]
I finished reviewing the rest of the commits. By and large this looks good. I deleted one comment from my previous review, but I think the others are still valid.

File: pkg/quota/gcp/limits.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+			Pages(ctx, func(page *serviceusage.ListConsumerQuotaMetricsResponse) error {
+				for _, qm := range page.Metrics {
+					for _, ql := range qm.ConsumerQuotaLimits {
+						for _, qlb := range ql.QuotaBuckets {
+							limit := record{
+								Service:  service[strings.LastIndex(service, "/")+1:],
+								Name:     ql.Metric,
+								Location: "global",
+							}
+							region, ok := qlb.Dimensions["region"]
```
[the remark this answers is not in the document]
Comment (abhinavdahiya) [reply pkg/quota/gcp/limits.go:39]: > https://cloud.google.com/service-usage/docs/reference/rest/v1beta1/services.consumerQuotaMetrics.limits#ConsumerQuotaLimit


quotaBuckets[] | object (QuotaBucket)                   Summary of the enforced quota buckets, organized by  quota dimension, ordered from least specific to most specific (for  example, the global default bucket, with no quota dimensions, will  always appear first).
-- | --


So the list is increasing order of specificity, and afaik there were at most 2 items either global or global and regional.

So for now i think this should be fine as-is, and if we do get more information that the assumption is wronf we can fix it as future work.

[the remark this answers is not in the document]
Comment (patrickdillon) [reply pkg/quota/gcp/limits.go:39]: Right. I saw this description, and they give the example of region, but not sure if there could be something completely unrelated like, e.g. `qlb.Dimensions["user"]` in which case a global record would be created. 

Makes sense to me that we fix it later if we see a problem.

**Pull request merged**